### DISCUSSION TITLE
Formatting merr/let

### DISCUSSION THREAD

**Discussion by arichiardi** [kkinnear/zprint#322] [Q&A] [answered]
Hi there :D Been a while.

I am trying to format the `merr/let` form [as per their documentation](
https://github.com/liquidz/merr?tab=readme-ov-file#concept) but I need some help.

Here is what I tried that unfortunately does not work:

```clojure
"merr/let" [:none {:next-inner
                     {:vector {:fn-format :binding}}}]
```

And here is an example and what I get:

```clojure
(ns core2.interpretation.zprint-test
  (:require
   [merr.core :as merr]))

(merr/let +err+
          [a (+ 1 1)
             b
             (do (merr/error {:msg "Oh noes!"}) 2)]
          (if +err+ (println "Error" (:msg +err+)) (- a b)))
```

Two problem I see:
  1. `:binding` does not apply to `b`
  2. The form outside binding are "not formatted" (do I need to use `:next-inner-restore` here?)

Any hint on how to improve that?

**Comment by kkinnear** [comment, marked as answer]
Hello!  Nice to hear from you.

Try this:
```
(czprint i322 {:parse-string? true :fn-map {"merr/let" [:guided {:guide [:element :element :element-binding-vec :newline :element-*]}]}})
(merr/let +err+ [a (+ 1 1)
                 b (do (merr/error {:msg "Oh noes!"}) 2)]
  (if +err+ (println "Error" (:msg +err+)) (- a b)))
```
I added the `guide` capability a couple of years ago.  Pretty complex to implement, but it allows me to do a lot of custom stuff.  If I had thought of it before I wrote zprint, I think I would have implemented all of zprint as guides.  I haven't documented it for use other than myself because it is pretty complex to document, and I doubt anyone would use it.  Generally, the guides are created by `option-fn` executables, where they are tailored to actually what is to be printed.  Sometimes things are simple enough that a fixed guide will solve the problem, as I think it does here. 

One of the limitations is that this guide will not ever print the `merr/let` all on one line.  But you probably don't care, since that would be a very short  and odd `let`. 

There may be a way to do this without using a guide, but this seemed the simpler way.  Let me know if it does or doesn't work in general when you try it.  Thanks!

**Reply by arichiardi** [reply]
Oh that's nice thank you!

Would that "guide" apply only to the element and then leave all the other elements untouched? For instance, the `(if +err+ ` would still follow my custom formatting rather than the "custom" one for the vector.

I hope I am making sense :D

**Comment by kkinnear** [comment]
The guide applies to all of the elements in the expression that starts with `merr/let`, but the only element that has its formatting actually altered is the vector, as a result of the `:element-binding-vector`.  Everything else is formatted as whatever is configured for that particular element.  Well, except for which line it is on, since the guide also affects when newlines are inserted, though additional newlines will be used if things get too wide.

**Reply by arichiardi** [reply]
As usual it took some time but I am not sure this is working here...

I have this code:

```clojure
(merr/let +err+ [result (validate-evaluation-result result) foo (clojure.pprint/pprint  'bar)]
            (if +err+
              (throw (errors/merr->exception +err+))
              result))
```
and I was expecting `foo` to go down no matter the width...

EDIT: re-reading, I don't think I specified that in the original message, apologies...

**Reply by kkinnear** [reply]
I can't reproduce that output, unfortunately.  Here is what I get:
```
zprint.core=> (print i322a)
(merr/let +err+ [result (validate-evaluation-result result) foo (clojure.pprint/pprint  'bar)]
            (if +err+
              (throw (errors/merr->exception +err+))
              result))
nil
zprint.core=> (czprint i322a {:parse-string? true :fn-map {"merr/let" [:guided {:guide [:element :element-best :element-binding-vec :newline :element-*]}]} :width 80})
(merr/let +err+ [result (validate-evaluation-result result)
                 foo (clojure.pprint/pprint 'bar)]
  (if +err+ (throw (errors/merr->exception +err+)) result))
nil
zprint.core=> (czprint i322a {:parse-string? true :fn-map {"merr/let" [:guided {:guide [:element :element-best :element-binding-vec :newline :element-*]}]} :width 800})
(merr/let +err+ [result (validate-evaluation-result result)
                 foo (clojure.pprint/pprint 'bar)]
  (if +err+ (throw (errors/merr->exception +err+)) result))
```
You can see I've tried with `:width 80` and just in case `:width 800`, and in both cases the binding vector is formatted as I (and I think you) expect and desire.

So, I'm thinking that something else in your configuration is impacting the formatting of the binding vector.

The formatting of the binding vector is controlled by the `:binding` key in the options map, and as I play with the various key-value pairs within `:binding`, I am unable to generate the particular output that you demonstrate.  I can *almost* generate that if I don't use the guide but rather `:fn-map {"mere/let" :arg2}`.  That does the binding vector as you have shown, but doesn't put the `(if +err+ ...)` in the place that yours ends up.

I'm also intrigued by the difference in how the `(if +err+ ...)` formats, where mine formats on a single line, and yours does not, despite there apparently being space for it to do so.

Anyway, I'm thinking that the best approach from here is for you to give me your entire configuration (your options map), and I'll see what I can figure out from that.  Should be interesting, I'm looking forward to it!

Thanks!

**Reply by arichiardi** [reply]
Hi @kkinnear and thanks for rubber ducking, indeed it was a configuration issue. I can now see `foo` down. I apologize for the confusion as it was quite late of Friday :smile: 

Thank you very much for the snippet as always. I'll start using guides more often now :+1:
